### PR TITLE
CTAD Deduction guides for TeamPolicy v2

### PR DESCRIPTION
**Pull request by nliber**
Seeing if adding an `Impl::TeamPolicyCommon` between `TeamPolicy` and `Impl::TeamPolicyInternal` will get around the compiler ICEing bugs.

### CODE REVIEW

**Comment by nliber** [comment]
retest this please

**Comment by nliber** [comment]
retest this please

File: core/src/Kokkos_ExecPolicy.hpp [hunk 1]
```diff
@@ -518,7 +518,229 @@ void team_policy_check_valid_storage_level_argument(int level);
...
+                   const Kokkos::AUTO_t&)
+      : internal_policy(league_size_request, team_size_request,
+                        Kokkos::AUTO()) {}
+
+  template <class... OtherProperties>
+  TeamPolicyCommon(const TeamPolicy<OtherProperties...> p)
+      : internal_policy(p) {
+    // Cannot call converting constructor in the member initializer list because
+    // it is not a direct base.
+    internal_policy::traits::operator=(p);
```
Comment (masterleinad) [inline core/src/Kokkos_ExecPolicy.hpp:630]: Why exactly is it necessary to call the copy assignment operator here? Why isn't `internal_policy(p)` enough?
I see that this is just copied from the existing implementation and introduced in https://github.com/kokkos/kokkos/pull/3379.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 2]
```diff
@@ -518,7 +518,229 @@ void team_policy_check_valid_storage_level_argument(int level);
...
+template <typename ES,
+          typename = std::enable_if_t<Kokkos::is_execution_space_v<ES>>>
+TeamPolicy(ES const&, int, Kokkos::AUTO_t const&, Kokkos::AUTO_t const&)
+    ->TeamPolicy<ES>;
+
+template <typename ES,
+          typename = std::enable_if_t<Kokkos::is_execution_space_v<ES>>>
+TeamPolicy(ES const&, int, int, Kokkos::AUTO_t const&)->TeamPolicy<ES>;
+
+#if 0
```
Comment (masterleinad) [inline core/src/Kokkos_ExecPolicy.hpp:743]: Can we delete this block now?

**Review by nmm0 — COMMENTED** [review]
Some initial comments

File: core/unit_test/TestTeamPolicyCTAD.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_Core_fwd.hpp"
```
Comment (dalg24) [inline core/unit_test/TestTeamPolicyCTAD.cpp:18]:
```suggestion
```

File: core/src/Kokkos_ExecPolicy.hpp [hunk 4]
```diff
@@ -600,120 +600,225 @@ void team_policy_check_valid_storage_level_argument(int level);
...
+  TeamPolicyCommon(const typename traits::execution_space& space_,
+                   int league_size_request, const Kokkos::AUTO_t&,
+                   const Kokkos::AUTO_t&)
       : internal_policy(space_, league_size_request, Kokkos::AUTO(),
                         Kokkos::AUTO()) {}
-  TeamPolicy(const typename traits::execution_space& space_,
-             int league_size_request, const int team_size_request,
-             const Kokkos::AUTO_t&)
+  TeamPolicyCommon(const typename traits::execution_space& space_,
+                   int league_size_request, const int team_size_request,
```
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/Kokkos_ExecPolicy.hpp:674]: Yeah.  There is all sorts of cleanup that can be done, such as passing `AUTO_t` by value, or if we look at more of the file, a bunch of `inline` that aren't necessary.  I don't mind a second PR after this is merged, but I don't want to make this one noisy.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 5]
```diff
@@ -600,120 +600,225 @@ void team_policy_check_valid_storage_level_argument(int level);
...
+                   int vector_length_request)
       : internal_policy(space_, league_size_request, team_size_request,
                         vector_length_request) {}
 
-  TeamPolicy(const typename traits::execution_space& space_,
-             int league_size_request, const Kokkos::AUTO_t&,
-             int vector_length_request = 1)
+  TeamPolicyCommon(const typename traits::execution_space& space_,
+                   int league_size_request, int team_size_request)
+      : internal_policy(space_, league_size_request, team_size_request, 1) {}
```
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/Kokkos_ExecPolicy.hpp:656]: Yes.  IIRC it had to do with inheriting constructors (`TeamPolicy` inherits `TeamPolicyCommon` constructors) and CTAD.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 6]
```diff
@@ -600,120 +600,225 @@ void team_policy_check_valid_storage_level_argument(int level);
...
-  using execution_policy = TeamPolicy<Properties...>;
+template <typename...>
+class TeamPolicyCommon;
+
+template <typename ExecSpace, typename... Properties>
+class TeamPolicyCommon<ExecSpace, Properties...>
+    : public TeamPolicyInternal<ExecSpace, Properties...> {
+  using internal_policy =
+      TeamPolicyInternal<typename PolicyTraits<Properties...>::execution_space,
+                         Properties...>;
```
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/Kokkos_ExecPolicy.hpp:637]: Done.  This was more about trying `void` as the first template parameter to `TeamPolicyCommon` (so changes anywhere else would be minimal), but as that didn't work and we are committing this, making this change is fine.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 7]
```diff
@@ -600,120 +600,225 @@ void team_policy_check_valid_storage_level_argument(int level);
...
+ * PolicyTraits<Properties...>::execution_space.
+ *
+ * Usage: TeamPolicy<Properties...> publicly derives from TeamPolicyCommon
+ * and inherits all constructors.
+ * Note: this cannot be enforced with a static_assert inside TeamPolicyCommon,
+ * as TeamPolicy is an incomplete type at that point.
+ *
+ * Note: We can't just use TeamPolicyInternal for this, because that
+ * is specialized for every ExecSpace (and therefore not common code).
  */
```
[the remark this answers is not in the document]
Comment (nliber) [reply core/src/Kokkos_ExecPolicy.hpp:622]: I took out the `\brief` marker, so it should no longer be doxygen syntax.

**Review by masterleinad — APPROVED** [review]
OK with me.

**Comment by dalg24** [comment]
Superseded by #7030

**Pull request closed without merging**